File: architecture-skills/design-patterns/examples/chain-of-responsibility/chain-of-responsibility.go
// Chain of Responsibility pattern in Go.
//
// Support ticket escalation. Go favors small interfaces or plain
// functions over linked handler objects: here the chain is a slice of
// handler funcs tried in order. net/http middleware wrapping
// (func(next http.Handler) http.Handler) is the same idea.

package main

import "fmt"

type Ticket struct {
	Subject  string
	Severity int // 1 (trivial) .. 4 (critical)
}

// HandlerFunc reports whether it resolved the ticket.
type HandlerFunc func(Ticket) bool

func supportLevel(name string, maxSeverity int) HandlerFunc {
	return func(t Ticket) bool {
		if t.Severity <= maxSeverity {
			fmt.Printf("%s resolved %q (severity %d)\n", name, t.Subject, t.Severity)
			return true
		}
		return false
	}
}

type Chain []HandlerFunc

func (c Chain) Dispatch(t Ticket) {
	for _, handle := range c {
		if handle(t) {
			return
		}
	}
	fmt.Printf("UNRESOLVED: %q (severity %d) -> escalate to engineering\n",
		t.Subject, t.Severity)
}

func main() {
	chain := Chain{
		supportLevel("Help bot", 1),
		supportLevel("Frontline agent", 2),
		supportLevel("Support manager", 3),
	}

	tickets := []Ticket{
		{"Reset my password", 1},
		{"Invoice shows wrong amount", 2},
		{"Account data corrupted", 3},
		{"Full region outage", 4},
	}

	for _, ticket := range tickets {
		chain.Dispatch(ticket)
	}
}

File: architecture-skills/design-patterns/examples/chain-of-responsibility/chain-of-responsibility.rs
// Chain of Responsibility pattern in Rust.
//
// Support ticket escalation. Instead of linked next-pointers (which
// fight the borrow checker), the idiomatic Rust form iterates a
// Vec<Box<dyn Handler>> until one handler claims the request.

struct Ticket {
    subject: &'static str,
    severity: u8, // 1 (trivial) .. 4 (critical)
}

trait Handler {
    /// Returns true if the ticket was resolved at this level.
    fn handle(&self, ticket: &Ticket) -> bool;
}

struct SupportLevel {
    name: &'static str,
    max_severity: u8,
}

impl Handler for SupportLevel {
    fn handle(&self, ticket: &Ticket) -> bool {
        if ticket.severity <= self.max_severity {
            println!(
                "{} resolved \"{}\" (severity {})",
                self.name, ticket.subject, ticket.severity
            );
            true
        } else {
            false
        }
    }
}

struct Chain {
    handlers: Vec<Box<dyn Handler>>,
}

impl Chain {
    fn dispatch(&self, ticket: &Ticket) {
        if !self.handlers.iter().any(|h| h.handle(ticket)) {
            println!(
                "UNRESOLVED: \"{}\" (severity {}) -> escalate to engineering",
                ticket.subject, ticket.severity
            );
        }
    }
}

fn main() {
    let chain = Chain {
        handlers: vec![
            Box::new(SupportLevel { name: "Help bot", max_severity: 1 }),
            Box::new(SupportLevel { name: "Frontline agent", max_severity: 2 }),
            Box::new(SupportLevel { name: "Support manager", max_severity: 3 }),
        ],
    };

    let tickets = [
        Ticket { subject: "Reset my password", severity: 1 },
        Ticket { subject: "Invoice shows wrong amount", severity: 2 },
        Ticket { subject: "Account data corrupted", severity: 3 },
        Ticket { subject: "Full region outage", severity: 4 },
    ];

    for ticket in &tickets {
        chain.dispatch(ticket);
    }
}
